### ISSUE TITLE
Error: Completion callback never invoked

### ISSUE THREAD

**Issue by mingyuelou1997** [mapbox/node-pre-gyp#367]
E:\TEST\sqlite3>npm install sqlite3

> sqlite3@4.0.0 install E:\TEST\sqlite3\node_modules\sqlite3
> node-pre-gyp install --fallback-to-build

node-pre-gyp ERR! Completion callback never invoked!
node-pre-gyp ERR! System Windows_NT 6.1.7601
node-pre-gyp ERR! command "D:\\nodejs\\node.exe" "E:\\TEST\\sqlite3\\node_module
s\\sqlite3\\node_modules\\node-pre-gyp\\bin\\node-pre-gyp" "install" "--fallback
-to-build"
node-pre-gyp ERR! cwd E:\TEST\sqlite3\node_modules\sqlite3
node-pre-gyp ERR! node -v v8.9.3
node-pre-gyp ERR! node-pre-gyp -v v0.9.0
node-pre-gyp ERR! This is a bug in `node-pre-gyp`.
node-pre-gyp ERR! Try to update node-pre-gyp and file an issue if it does not he
lp:
node-pre-gyp ERR!     <https://github.com/mapbox/node-pre-gyp/issues>
npm ERR! code ENOSYS
npm ERR! errno ENOSYS
npm ERR! syscall kill
npm ERR! kill ENOSYS

**Comment by netbuffer** [comment]
face the same issue

**Comment by springmeyer** [comment]
> Completion callback never invoked

@mapsam also saw this recently on travis. I'm going to rename the title of this ticket `Completion callback never invoked`.

**Comment by mapsam** [comment]
Indeed - here's an OSX build with node-pre-gyp tests failing (irregularly) with the same error: https://travis-ci.org/mapbox/node-pre-gyp/jobs/372259045

**Comment by skyh24** [comment]
how to solve this？？？？

**Comment by TaikiAkita** [comment]
Really sucks when I'm building node-webrtc...
Please solve this FV*King problem immediately!!!!

**Comment by IgorMaineti** [comment]
Could someone look into this issue please? I'm facing the same problem

**Comment by Rosso84** [comment]
has anyone fixed this issue yet?? Mine always works for a short period of time then all of a sudden this issue shows up.. After i gave up on wondows i dualbooted with an ubuntu and installed it there then it fixed  my problem.. Then it worked for a short period and then the problem came back.. I immediately installed it on a third OS , this time on a mac but the same issue showed up fro the beginning.. I have a feeling Node is simply broken... my education  went to hell because of this crappy stuff.

**Comment by eraldinho** [comment]
Hello, I had the same problem, I disabled my antivirus (kaspersky) and it ran like a charm. Hope it can help.

**Comment by paulmbw** [comment]
thank you @eraldinho !!!!

**Comment by nightwolf13** [comment]
Is there any other way to solve it?.. I can't disable my antivirus, because it's enterprise..

**Comment by eraldinho** [comment]
I don't know, sorry

**Comment by D4you208** [comment]
@nightwolf13  Do you find any solution ? I am getting same error.

**Comment by nightwolf13** [comment]
@D4you208 I had this issue when using firebase cloud functions.. (it was used by firebase-admin i think).. So i solved it by "npm install firebase-admin --force".. But this solution could work only for cloud functions, because it's executed on the server and i think there isn't such problem.. So all i need to install it locally to pass lint check before deploy.. (i think it will not work locally)

**Comment by pgn-vole** [comment]
Same error here as well.

**Comment by rathodamol** [comment]
someone pls help!

**Comment by emilwidlund** [comment]
This is really annoying.

**Comment by rathodamol** [comment]
npm install -g node-pre-gyp       ...(this will install v0.9.0)
npm update node-pre-gyp   ..(this will update it to 0.10.0)

or alternatively 
npm install -g node-pre-gyp@0.10.0

**Comment by mapsam** [comment]
Thanks to #377 we're able to get more clarity on this issue - seems related to socket timeouts from s3 (perhaps related to the usage of needle instead of request.js). Working on debugging over at https://github.com/mapbox/node-pre-gyp/pull/394

**Comment by RichardSilveira** [comment]
For those are still facing this issue, give a try on: https://stackoverflow.com/questions/39674390/error-with-npm-install-a-pre-gyp-error

**Comment by mapsam** [comment]
Hi all, I'm setting this down for a bit as I'll be away from a computer for the next few weeks. That said, I want to leave notes of patterns I've seen for anyone who may want to pick this up and keep debugging.

* I haven't been able to recreate this locally at all, so debugging is a slow process on travis builds right now
* Travis recreates this regularly, and right now the main lead is that we're running into socket timeouts with s3 with both the `publish` and `install` commands. See more debugging information over at #394 - increasing open/read/connection timeouts with needle hasn't led to any improvement
* I've disabled "PR push" builds on travis to avoid race condition failures, which are noted in #373 
* This appears to be semi-related to proxies and global/lack of global node-pre-gyp/node-gyp installations, so if anyone is willing to share their system info and stack traces, it'd be welcome so we can start seeing more patterns here
* I've seen this happen on all node versions we're testing - 6/8/9/10

Next steps I'd suggest/attempt:

* revert back to using request.js (instead of needle) to see if the problem persists
* Create a separate branch/repository test case that attempts to do similar installs/puts with s3 HTTP endpoints or the node.js aws-sdk module and attempt to recreate on travis-ci
* soliciting for stack traces of others hitting this error as stated above (please provide a link to a gist if possible to keep this issue concise!)
* Try installing again with d1730cc8ce5f6bc4b5e0314d25ee9c71767d04f5 to see if you get new socket errors instead of "callback never invoked". This will at least confirm this issue is actually related to socket timeouts as we see in #394. I've published `0.10.1-dev.1` of node-pre-gyp to test this out.

**Comment by zixia** [comment]
I run into this issue today in my Mac Pro with node v8.

Is there a workaround to make it work?

**Comment by springmeyer** [comment]
Thanks for the detailed report @zixia 

> I believe it's a network issue on my case. We should have more clear error message for this kind of situation.

If you upgrade (or ask the grpc developers to upgrade) node-pre-gyp to v0.10.2 I think the error message will improve since v0.10.2 includes https://github.com/mapbox/node-pre-gyp/pull/377.

A potential root cause would be network failure or potentially (if you are behind a proxy with one network) #400

**Comment by kalmecak** [comment]
I had same error when trying to install firebase-admin (5.12.1) on Ubuntu 18.04 I solved it following these steps:

*Clone grpc directly on node_modules folder*
`$ git clone git@github.com:grpc/grpc.git`

*Enter to node_modules/grpc and:*
`$ npm install --build-from-source`

After that I install firebase-admin and works but when installation ends npm throws an error:
`grpc: Appears to be a git repo or submodule`
it means that npm can not erase grpc folder.

Hope it helps, it is no the better way but works.

**Comment by springmeyer** [comment]
This should be fixed by node-pre-gyp@0.10.1: https://github.com/mapbox/node-pre-gyp/blob/master/CHANGELOG.md#0101. If anyone hits this again, please create a new issue.

**Comment by amreladawy** [comment]
For me, it was a network issue. switched to mobile hotspot and it worked

**Comment by molerat619** [comment]
The hotspot didn't help me. And how would that help on a production server that you can only ssh in?

**Comment by DevinXian** [comment]
@rathodamol It works for me

**Comment by ImTheC** [comment]
The error I received when running `npm i` on a network with proxy:

```
node-pre-gyp ERR! Completion callback never invoked! 
node-pre-gyp ERR! System Darwin 18.2.0
node-pre-gyp ERR! command "/usr/local/bin/node" "<XXX>/node_modules/fsevents/node_modules/node-pre-gyp/bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd <XXX>/node_modules/fsevents
node-pre-gyp ERR! node -v v9.9.0
node-pre-gyp ERR! node-pre-gyp -v v0.10.0
node-pre-gyp ERR! This is a bug in `node-pre-gyp`.
```

But thanks to everyone's comments, I switched to a network without proxy and the error went away. Then I got this message:

```
> fsevents@1.2.4 install <XXX>/node_modules/fsevents
> node install

[fsevents] Success: "<XXX>/node_modules/fsevents/lib/binding/Release/node-v59-darwin-x64/fse.node" is installed via remote
```

**Comment by msbadar** [comment]
Try with `yarn` instead of npm. it works form me.

**Comment by stefanmatar** [comment]
For me this solved by connecting to the internet. I didn't have an internet connection when building and it seemed like it needed some dependencies.

**Comment by aconchillo** [comment]
In case it helps to someone, mine was a permissions issue, see https://github.com/mapbox/node-pre-gyp/issues/433#issuecomment-471028255

**Comment by springmeyer** [comment]
This should be fixed now as long as your package-lock.json does not try to install `minipass@2.8.4`: see https://github.com/mapbox/node-pre-gyp/issues/477#issuecomment-534630617

**Comment by Andersonfrfilho** [comment]
was having the same problem, uninstalled Node 12 and installed Node version 10.19 and it worked for me
```https://nodejs.org/dist/latest-v10.x/```

**Comment by adityak74** [comment]
It was a network issue for me and connecting to a US VPN server solved the issue.

**Comment by csgka1** [comment]
I met this issue when i install grpc-tools@0.15.0 **globally**.  I tried to install `request` as the default http lib of `node-pre-gyp` to replace the original `needle`. But it didn't work. Then I aborted the installing progress manually after the npm downloading and before the node-pre-gyp throwing error. Then I deleted the directory of `grpc-tools`. Finally I rerun `npm install grpc-tools -g`, and it worked, no error. yes, it worked, and I don't know why.
